sdk/clientcore/http-netty4/src/main/java/io/clientcore/http/netty4/implementation/CoreHttpProxyHandler.java

```diff
@@ -0,0 +1,256 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package io.clientcore.http.netty4.implementation;
+
+import io.clientcore.core.http.models.HttpHeaderName;
+import io.clientcore.core.http.models.ProxyOptions;
+import io.clientcore.core.instrumentation.logging.ClientLogger;
+import io.clientcore.core.utils.AuthUtils;
+import io.clientcore.core.utils.AuthenticateChallenge;
+import io.clientcore.core.utils.CoreUtils;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.DefaultFullHttpRequest;
+import io.netty.handler.codec.http.FullHttpRequest;
+import io.netty.handler.codec.http.HttpClientCodec;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpUtil;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.handler.proxy.HttpProxyHandler;
+import io.netty.handler.proxy.ProxyConnectException;
+import io.netty.handler.proxy.ProxyHandler;
+
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static io.clientcore.core.utils.AuthUtils.parseAuthenticationOrAuthorizationHeader;
+import static io.clientcore.http.netty4.implementation.NettyUtility.createCodec;
+
+/**
+ * This class handles authorizing requests being sent through a proxy which require authentication.
+ */
+public final class CoreHttpProxyHandler extends ProxyHandler {
```

Comment:
Should something named `Core*` be in the `netty4` package?

Reply:
I really wasn't sure what to call this as Netty has an `HttpProxyHandler` too, and by default a lot of logging around `Channel` and `ChannelPipeline` use the class name, and that could be confusing.

I'm happy to rename this. Maybe `Netty4HttpProxyHandler`?

Reply:
`Netty4HttpProxyHandler` is great.